build-tools-internal/src/integTest/java/org/elasticsearch/gradle/internal/SymbolicLinkPreservingTarIT.java

```diff
@@ -0,0 +1,154 @@
+import org.gradle.api.GradleException;
+import org.gradle.testkit.runner.GradleRunner;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static org.hamcrest.CoreMatchers.anyOf;
+import static org.hamcrest.CoreMatchers.equalTo;
+
+public class SymbolicLinkPreservingTarIT extends GradleIntegrationTestCase {
+
+    @Override
+    public String projectName() {
+        return "symbolic-link-preserving-tar";
+    }
+
+    @Rule
+    public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Before
+    public void before() throws IOException {
+        final Path realFolder = temporaryFolder.getRoot().toPath().resolve("real-folder");
+        Files.createDirectory(realFolder);
+        Files.createFile(realFolder.resolve("file"));
+        Files.createSymbolicLink(realFolder.resolve("link-to-file"), Paths.get("./file"));
+        final Path linkInFolder = temporaryFolder.getRoot().toPath().resolve("link-in-folder");
+        Files.createDirectory(linkInFolder);
+        Files.createSymbolicLink(linkInFolder.resolve("link-to-file"), Paths.get("../real-folder/file"));
+        final Path linkToRealFolder = temporaryFolder.getRoot().toPath().resolve("link-to-real-folder");
+        Files.createSymbolicLink(linkToRealFolder, Paths.get("./real-folder"));
+    }
+
+    public void testBZip2Tar() throws IOException {
+        runBuild("buildBZip2Tar", true);
+        assertTar(".bz2", BZip2CompressorInputStream::new, true);
+    }
+
+    public void testBZip2TarDoNotPreserveFileTimestamps() throws IOException {
+        runBuild("buildBZip2Tar", false);
+        assertTar(".bz2", BZip2CompressorInputStream::new, false);
+    }
+
+    public void testGZipTar() throws IOException {
+        runBuild("buildGZipTar", true);
+        assertTar(".gz", GzipCompressorInputStream::new, true);
+    }
+
+    public void testGZipTarDoNotPreserveFileTimestamps() throws IOException {
+        runBuild("buildGZipTar", false);
+        assertTar(".gz", GzipCompressorInputStream::new, false);
+    }
+
+    public void testTar() throws IOException {
+        runBuild("buildTar", true);
+        assertTar("", fis -> fis, true);
+    }
+
+    public void testTarDoNotPreserveFileTimestamps() throws IOException {
+        runBuild("buildTar", false);
+        assertTar("", fis -> fis, false);
+    }
+
+    interface FileInputStreamWrapper {
+        InputStream apply(FileInputStream fis) throws IOException;
+    }
+
+    private void assertTar(final String extension, final FileInputStreamWrapper wrapper, boolean preserveFileTimestamps)
+        throws IOException {
+        try (TarArchiveInputStream tar = new TarArchiveInputStream(wrapper.apply(new FileInputStream(getOutputFile(extension))))) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileInputStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *FileInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).